tests/test_args_specs.py

```diff
@@ -0,0 +1,335 @@
+                    # is too complex to describe this way.
+                    has_rows = 'rows' in arg
+                    has_cols = 'columns' in arg
+                    # should not have both
+                    self.assertTrue(not (has_rows and has_cols))
+
+                    if has_cols or has_rows:
+                        direction = 'rows' if has_rows else 'columns'
+                        headers = arg[direction]
+                        self.assertTrue(isinstance(headers, dict))
+
+                        for header in headers:
+                            self.assertTrue(isinstance(header, str))
+                            if '(' in header or '[' in header:
+                                print(f"Warning: header {header} in {name}")
+                            self.validate(
+                                headers[header],
+                                f'{name}.{direction}.{header}',
+                                valid_types=valid_nested_types['csv'])
+
+                        attrs.discard('rows')
+                        attrs.discard('columns')
+
+                elif t == 'directory':
+                    # directory type should have a contents property that maps each
+                    # expected path name/pattern within the directory to a nested
+                    # type dictionary describing the data at that filepath
+                    self.assertTrue('contents' in arg)
+                    self.assertTrue(isinstance(arg['contents'], dict))
+                    for path in arg['contents']:
+                        self.assertTrue(isinstance(path, str))
+                        if '(' in path or '[' in path:
+                            print(f"Warning: header {path} in {name}")
+                        self.validate(
+                            arg['contents'][path],
+                            f'{name}.contents.{path}',
+                            valid_types=valid_nested_types['directory'])
+                    attrs.remove('contents')
+
+            # iterate over the remaining attributes
+            # type-specific ones have been removed by this point
+            if 'name' in attrs:
+                self.assertTrue(isinstance(arg['name'], str))
+                attrs.remove('name')
+            if 'about' in attrs:
+                self.assertTrue(isinstance(arg['about'], str))
+                attrs.remove('about')
+            if 'required' in attrs:
+                # required value may be True, False, or a string that can be
+                # parsed as a python statement that evaluates to True or False
+                self.assertTrue(isinstance(arg['required'], bool) or
+                                isinstance(arg['required'], str))
+                attrs.remove('required')
+            if 'type' in attrs:
+                self.assertTrue(isinstance(arg['type'], str) or
+                                isinstance(arg['type'], set))
```

Comment:
There are a few:
- globio's infrastructure directory input can contain rasters and/or vectors
- the "path" column in the rec model predictor table can have rasters and/or vectors
- the HRA habitat stressor info table "path" column can have rasters and/or vectors
- the HRA criteria scores table can have cells with rasters, vectors, and/or integers (but this table structure isn't represented in the spec because it's too complicated)